src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -3120,6 +3120,16 @@ StorageReplicatedMergeTree::CreateMergeEntryResult StorageReplicatedMergeTree::c
     int32_t log_version,
     MergeType merge_type)
 {
+    /// It is possible that intersected part had been added while fetching replication queue
+    /// (i.e. some replica already created entry that includes different set of parts).
+    ///
+    /// NOTE: since we check log_version race should not be possible.
+    if (queue.isVirtualPart(merged_info, merged_name))
+    {
+        throw Exception(ErrorCodes::DUPLICATE_DATA_PART,
+            "There is already queue entry that intersects {} part", merged_name);
+    }
+
```

Comment:
AFAIK if there is a covering part in the queue, then either `ReplicatedMergeTreeMergePredicate::canMergeTwoParts` returned `false` or zookeeper request will fail due to log version mismatch, so this change looks redundant. Could you please explain in more details where is the race condition and why this check is supposed to fix it?

Reply:
Yep, seems you are right, and this change is redundant.